File: apple/Sources/ClipCore/WSMessages.swift
import Foundation

public struct OnlineDevice: Codable, Sendable, Equatable, Hashable, Identifiable {
    public var deviceId: String
    public var name: String
    public var platform: String

    public var id: String { deviceId }

    public init(deviceId: String, name: String, platform: String) {
        self.deviceId = deviceId
        self.name = name
        self.platform = platform
    }
}

public struct WelcomeMessage: Codable, Sendable, Equatable {
    public var protocolVersion: Int
    public var serverVersion: String
    public var serverId: String
    public var serverTime: Date
    public var deviceId: String
    public var currentSeq: Int64
    public var stateRev: Int64
    public var onlineDevices: [OnlineDevice]

    public init(protocolVersion: Int = 1, serverVersion: String = "test", serverId: String, serverTime: Date, deviceId: String, currentSeq: Int64, stateRev: Int64, onlineDevices: [OnlineDevice] = []) {
        self.protocolVersion = protocolVersion
        self.serverVersion = serverVersion
        self.serverId = serverId
        self.serverTime = serverTime
        self.deviceId = deviceId
        self.currentSeq = currentSeq
        self.stateRev = stateRev
        self.onlineDevices = onlineDevices
    }
}

public struct PresenceMessage: Codable, Sendable, Equatable {
    public var deviceId: String
    public var name: String
    public var platform: String
    public var online: Bool
}

public struct ClipMessage: Codable, Sendable, Equatable {
    public var item: ItemHeader
}

public struct ClipDeletedMessage: Codable, Sendable, Equatable {
    public var ids: [String]
    public var reason: String
    public var stateRev: Int64

    public init(ids: [String], reason: String, stateRev: Int64) {
        self.ids = ids
        self.reason = reason
        self.stateRev = stateRev
    }
}

public struct ClipPinnedMessage: Codable, Sendable, Equatable {
    public var id: String
    public var pinned: Bool
    public var stateRev: Int64

    public init(id: String, pinned: Bool, stateRev: Int64) {
        self.id = id
        self.pinned = pinned
        self.stateRev = stateRev
    }
}

public struct StorageWarningMessage: Codable, Sendable, Equatable {
    public var active: Bool
    public var reason: String
    public var freeDiskBytes: Int64
    public var minFreeDiskBytes: Int64
}

public struct WSErrorMessage: Codable, Sendable, Equatable {
    public var code: String
    public var message: String
    public var details: JSONValue?
}

public enum ServerMessage: Sendable, Equatable {
    case welcome(WelcomeMessage)
    case presence(PresenceMessage)
    case devicesChanged
    case clip(ItemHeader)
    case clipDeleted(ClipDeletedMessage)
    case clipPinned(ClipPinnedMessage)
    case storageWarning(StorageWarningMessage)
    case ping(Int64)
    case pong(Int64)
    case error(WSErrorMessage)
    case unknown(String)

    private struct TypeProbe: Decodable {
        var type: String
    }

    private struct TsBody: Decodable {
        var ts: Int64
    }

    public static func decode(_ data: Data) throws -> ServerMessage {
        let d = JSONCoding.decoder()
        let type = try d.decode(TypeProbe.self, from: data).type
        switch type {
        case "welcome": return .welcome(try d.decode(WelcomeMessage.self, from: data))
        case "presence": return .presence(try d.decode(PresenceMessage.self, from: data))
        case "devices_changed": return .devicesChanged
        case "clip": return .clip(try d.decode(ClipMessage.self, from: data).item)
        case "clip_deleted": return .clipDeleted(try d.decode(ClipDeletedMessage.self, from: data))
        case "clip_pinned": return .clipPinned(try d.decode(ClipPinnedMessage.self, from: data))
        case "storage_warning": return .storageWarning(try d.decode(StorageWarningMessage.self, from: data))
        case "ping": return .ping(try d.decode(TsBody.self, from: data).ts)
        case "pong": return .pong(try d.decode(TsBody.self, from: data).ts)
        case "error": return .error(try d.decode(WSErrorMessage.self, from: data))
        default: return .unknown(type)
        }
    }

    public static func decode(_ text: String) throws -> ServerMessage {
        try decode(Data(text.utf8))
    }
}

public enum ClientMessage: Sendable, Equatable {
    case hello(deviceId: String, lastSeq: Int64, appVersion: String)
    case ping(Int64)
    case pong(Int64)

    private struct Hello: Encodable {
        var type = "hello"
        var protocolVersion: Int
        var deviceId: String
        var lastSeq: Int64
        var appVersion: String
        var platform: String
    }

    private struct Ts: Encodable {
        var type: String
        var ts: Int64
    }

    public func encoded() throws -> Data {
        let e = JSONCoding.encoder()
        switch self {
        case .hello(let deviceId, let lastSeq, let appVersion):
            return try e.encode(Hello(protocolVersion: Proto.version, deviceId: deviceId, lastSeq: lastSeq, appVersion: appVersion, platform: Proto.platform))
        case .ping(let ts):
            return try e.encode(Ts(type: "ping", ts: ts))
        case .pong(let ts):
            return try e.encode(Ts(type: "pong", ts: ts))
        }
    }

    public func text() throws -> String {
        String(decoding: try encoded(), as: UTF8.self)
    }
}

File: android/core/src/main/kotlin/dev/yikz/clipboard/core/WsMessages.kt
package dev.yikz.clipboard.core

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonObject

sealed interface WsMessage

@Serializable
data class HelloMsg(
    val type: String = "hello",
    val protocolVersion: Int = Protocol.VERSION,
    val deviceId: String,
    val lastSeq: Long,
    val appVersion: String,
    val platform: String = Protocol.PLATFORM_ANDROID,
) : WsMessage

@Serializable
data class OnlineDevice(val deviceId: String, val name: String = "", val platform: String = "")

@Serializable
data class WelcomeMsg(
    val type: String = "welcome",
    val protocolVersion: Int,
    val serverVersion: String = "",
    val serverId: String,
    val serverTime: String,
    val deviceId: String,
    val currentSeq: Long,
    val stateRev: Long,
    val onlineDevices: List<OnlineDevice> = emptyList(),
) : WsMessage

@Serializable
data class PresenceMsg(
    val type: String = "presence",
    val deviceId: String,
    val name: String = "",
    val platform: String = "",
    val online: Boolean,
) : WsMessage

@Serializable
data class DevicesChangedMsg(val type: String = "devices_changed") : WsMessage

@Serializable
data class ClipMsg(val type: String = "clip", val item: ItemHeader) : WsMessage

@Serializable
data class ClipDeletedMsg(
    val type: String = "clip_deleted",
    val ids: List<String>,
    val reason: String = "",
    val stateRev: Long,
) : WsMessage

@Serializable
data class ClipPinnedMsg(
    val type: String = "clip_pinned",
    val id: String,
    val pinned: Boolean,
    val stateRev: Long,
) : WsMessage

@Serializable
data class StorageWarningMsg(
    val type: String = "storage_warning",
    val active: Boolean,
    val reason: String = "",
    val freeDiskBytes: Long = 0,
    val minFreeDiskBytes: Long = 0,
) : WsMessage

@Serializable
data class ReleaseAvailableMsg(val type: String = "release_available", val version: String) : WsMessage

@Serializable
data class PingMsg(val type: String = "ping", val ts: Long) : WsMessage

@Serializable
data class PongMsg(val type: String = "pong", val ts: Long) : WsMessage

@Serializable
data class ErrorMsg(
    val type: String = "error",
    val code: String,
    val message: String = "",
    val details: JsonObject? = null,
) : WsMessage

data class UnknownMsg(val type: String) : WsMessage

data class InvalidMsg(val reason: String) : WsMessage

object WsCodec {
    fun decode(text: String): WsMessage {
        val obj = try {
            ProtocolJson.parseToJsonElement(text).jsonObject
        } catch (e: Exception) {
            return InvalidMsg("not a JSON object")
        }
        val typeElement = obj["type"] as? JsonPrimitive
        if (typeElement == null || !typeElement.isString) return InvalidMsg("missing type")
        val type = typeElement.content
        return try {
            when (type) {
                "welcome" -> ProtocolJson.decodeFromJsonElement(WelcomeMsg.serializer(), obj)
                "presence" -> ProtocolJson.decodeFromJsonElement(PresenceMsg.serializer(), obj)
                "devices_changed" -> DevicesChangedMsg()
                "clip" -> ProtocolJson.decodeFromJsonElement(ClipMsg.serializer(), obj)
                "clip_deleted" -> ProtocolJson.decodeFromJsonElement(ClipDeletedMsg.serializer(), obj)
                "clip_pinned" -> ProtocolJson.decodeFromJsonElement(ClipPinnedMsg.serializer(), obj)
                "storage_warning" -> ProtocolJson.decodeFromJsonElement(StorageWarningMsg.serializer(), obj)
                "release_available" -> ProtocolJson.decodeFromJsonElement(ReleaseAvailableMsg.serializer(), obj)
                "ping" -> ProtocolJson.decodeFromJsonElement(PingMsg.serializer(), obj)
                "pong" -> ProtocolJson.decodeFromJsonElement(PongMsg.serializer(), obj)
                "error" -> ProtocolJson.decodeFromJsonElement(ErrorMsg.serializer(), obj)
                "hello" -> ProtocolJson.decodeFromJsonElement(HelloMsg.serializer(), obj)
                else -> UnknownMsg(type)
            }
        } catch (e: Exception) {
            InvalidMsg("malformed $type: ${e.message}")
        }
    }

    fun encode(message: WsMessage): String = when (message) {
        is HelloMsg -> ProtocolJson.encodeToString(HelloMsg.serializer(), message)
        is PingMsg -> ProtocolJson.encodeToString(PingMsg.serializer(), message)
        is PongMsg -> ProtocolJson.encodeToString(PongMsg.serializer(), message)
        is WelcomeMsg -> ProtocolJson.encodeToString(WelcomeMsg.serializer(), message)
        is PresenceMsg -> ProtocolJson.encodeToString(PresenceMsg.serializer(), message)
        is DevicesChangedMsg -> ProtocolJson.encodeToString(DevicesChangedMsg.serializer(), message)
        is ClipMsg -> ProtocolJson.encodeToString(ClipMsg.serializer(), message)
        is ClipDeletedMsg -> ProtocolJson.encodeToString(ClipDeletedMsg.serializer(), message)
        is ClipPinnedMsg -> ProtocolJson.encodeToString(ClipPinnedMsg.serializer(), message)
        is StorageWarningMsg -> ProtocolJson.encodeToString(StorageWarningMsg.serializer(), message)
        is ErrorMsg -> ProtocolJson.encodeToString(ErrorMsg.serializer(), message)
        is ReleaseAvailableMsg -> ProtocolJson.encodeToString(ReleaseAvailableMsg.serializer(), message)
        is UnknownMsg, is InvalidMsg -> throw IllegalArgumentException("cannot encode $message")
    }
}

object CloseCodes {
    const val NORMAL = 1000
    const val GOING_AWAY = 1001
    const val TOO_LARGE = 1009
    const val UNAUTHORIZED = 4001
    const val TOO_MANY_CONNECTIONS = 4002
    const val VERSION_UNSUPPORTED = 4003
    const val HELLO_REQUIRED = 4004
    const val HEARTBEAT_TIMEOUT = 4005
}
